userd/launcher.go

```diff
@@ -93,3 +102,78 @@ func (s *Launcher) OpenURL(addr string) *dbus.Error {
 
 	return nil
 }
+
+func fdToFilename(fd int) (string, error) {
+	flags, err := sys.FcntlGetFl(fd)
+	if err != nil {
+		return "", err
+	}
+	// File descriptors opened with O_PATH do not imply access to
+	// the file in question.
+	if flags&sys.O_PATH != 0 {
+		return "", fmt.Errorf("bad file descriptor")
+	}
+
+	// Determine the file name associated with the passed file descriptor.
+	filename, err := os.Readlink(fmt.Sprintf("/proc/self/fd/%d", fd))
+	if err != nil {
+		return "", err
+	}
+
+	var fdStat, fileStat syscall.Stat_t
+	if err := syscall.Fstat(fd, &fdStat); err != nil {
+		return "", err
+	}
+	if err := syscall.Stat(filename, &fileStat); err != nil {
+		return "", err
+	}
+
+	// Sanity check to ensure we've got the right file
+	if fdStat.Dev != fileStat.Dev || fdStat.Ino != fileStat.Ino {
+		return "", fmt.Errorf("could not determine file name")
+	}
+
+	fileType := fileStat.Mode & syscall.S_IFMT
+	if fileType != syscall.S_IFREG && fileType != syscall.S_IFDIR {
+		return "", fmt.Errorf("can only open regular files and directories")
+	}
+
+	return filename, nil
+}
+
+func (s *Launcher) OpenFile(clientFd dbus.UnixFD, sender dbus.Sender) *dbus.Error {
+	// godbus transfers ownership of this file descriptor to us
+	fd := int(clientFd)
+	defer syscall.Close(fd)
+
+	filename, err := fdToFilename(fd)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	snap, err := snapFromSender(s.conn, sender)
```

Comment:
I think a better fix would be to change the code to call `GetConnectionCredentials` instead of GetConnectionUnixProcessID`, and look at `LINUX_SECURITY_LABEL` in the returned dictionary. This avoids the PID reuse problem, since the bus unique names aren't reused.

Closing the `dbus.Conn` seems incorrect though: it represents connection to the bus, rather than a connection to another peer on the bus.

Reply:
I considered solely using GetConnectionCredentials() instead of GetConnectionUnixProcessID() with a cgroup lookup, but we would need the cgroup lookup for cross-distro since there won't be a label on systems where AppArmor isn't enabled (that's why we have the cgroup check in the first place). Sure, we could check the confinement level and use GetConnectionCredentials if we expect to have a label, otherwise GetConnectionUnixProcessID, but I liked how calling GetConnectionUnixProcessID and detecting if the connection closed fixed both and can therefore ensure the proven check regardless.

Really though, I wanted to call something along the lines of "is connection still connected" to see if the client died, but I couldn't find in godbus a call to do that, so just reused what snapd already implemented. I also realized that the unique name isn't reused, but 'sender' isn't updated in what I outlined because process B didn't reconnect to the bus, it just reused the PID, so we couldn't check the sender's connection name (indeed, within the call to snapFromSender it wouldn't ever change). IIUC dbus-daemon is supposed to send a signal when the clients disconnects, so what is probably most correct would be to detect this and terminate OpenFile if it did, but I didn't readily see how to do that in godbus. I'm certainly open to other ideas, but one more call to detect for disconnection was cheap.

As for closing dbus.Conn, I was thinking that it represented the connection between the server and the client for this unique connection name. Perhaps that is wrong, but I can say that when I tested this, the server didn't stop listening for new connections, which seemed to reaffirm my thinking. The call isn't strictly needed though since we return a DBus error instead of opening the file.

Reply:
@mvo5 - this is the bit that needs to be addressed before committing (please see whole comment)

Reply:
I've updated `snapFromSender` to perform a `NameHasOwner` call on the sender unique bus name after we've looked up the snap from the process ID.  This will fail if the sender has disconnected from the bus, which would happen if the process A exits.

As mentioned earlier, we're not going to close the connection to the bus, because it will leave a zombie userd around disconnected from the session bus.  I suspect that things appeared to work for @jdstrand because a new instance of userd was spawned by D-Bus service activation.